<a href="https://colab.research.google.com/github/Thippawan72/BSC_DPDM23/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# กลุ่ม ช่วยด้วย
### สมาชิก
        - นางสาวทิพย์วัลย์ สุโพธิ์                 รหัส 643020062-6        สไลด์ + present
        - นางสาวศดานันท์ ทรัพย์มีมหาศาล    รหัส 643021272-0        สไลด์ + present
        - นางสาวสุชานาถ พิลาภ                  รหัส 643021276-2        สไลด์ + present
        - นางสาวอิงอร พลพาล                    รหัส 643021279-6        Coding + present

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Mounted at /content/drive


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
!pip install apyori  #โหลดเพื่อทำ Association Rule

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=15609fc126e6c1b947d0e47a5d08a8d481b17b4e21ea6b3aef0bae22290fc750
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
from apyori import apriori

# HW1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลาตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน (groupby ประเทศกับเดือน)

### จัดการวันที่

In [ ]:
import datetime

In [ ]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [ ]:
# แปลงคอลัมน์ให้เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])


In [ ]:
#เปลี่ยนรูปแบบให้เป็น เดือน / วัน / ปี
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%y')

In [ ]:
# สร้างคอลัมน์ 'Date' โดยดึงข้อมูลเฉพาะวันที่
data['Date'] = data['InvoiceDate'].dt.date
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
...                    ...        ...         ...             ...         ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09  

[541909 rows x 9 columns]

In [ ]:
data['month'] = data['InvoiceDate'].dt.month

In [ ]:
data['day'] = data['InvoiceDate'].dt.day

## ฟังก์ชันที่หาว่าสินค้านั้นคืออะไร



In [ ]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

# ข้อ 1 หาข้อมูลของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)

In [ ]:
# ดูช่วงเวลาคริสต์มาสกับปีใหม่
Christm_Newy = data[data['month']==12]
Christm_Newy

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
541904     12    9  
541905     12    9  
541906     12    9  
541907     12    9  
541908     12    9  

[68006 rows x 11 columns]

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

trans_Christm_Newy = Christm_Newy.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในช่วงเวลาคริสต์มาสกับปีใหม่ โดยมีค่า min support คือ 8%
frequentItemsets= list(apriori(trans_Christm_Newy,min_support = 0.08))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),

 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),

 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]





In [ ]:
# จำนวนสินค้า
len(frequentItemsets)

3

In [ ]:
# หาสินค้าที่ถูกซื้อซ้ำจาก trans_Christm_Newy โดยกำหนดค่า min_sup = 0.045  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.045

frequentItem_CN = list(apriori(trans_Christm_Newy,min_support = min_sup))
for i in frequentItem_CN:
    if i[0].__len__()>1:
        print('yes')
        print(i)


yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0),

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22910') | (data['StockCode']=='22086')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

# ข้อ 2 หาข้อมูลของประเทศในทวีป Europe

In [ ]:
#แยกประเทศที่อยู่ในทวีปยุโรป

Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [ ]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
536964 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536965 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536966 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536967 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536968 2011-12-08 10:26:00      40.00     12358.0         Austria  2011-12-08   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
536964     12    8  
536965     12    8  
536966     12    8  
536967     12    8  
536968     12    8  

[538358 rows x 11 columns]

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในทวีปยุโรป โดยมีค่า min support คือ 5%

frequentItemsetsEU = list(apriori(transactionsEU,min_support = 0.05))
frequentItemsetsEU

[RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20727'}), support=0.05009521588745094, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20727'}), confidence=0.05009521588745094, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.05149430647856671, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.05149430647856671, lift=1.0)]),
 RelationRecord(items=frozenset({'22197'}), support=0.05584703276203801, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22197'}), confidence=0.05584703276203801, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.050600443045353855, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({

In [ ]:
# จำนวนสินค้า
len(frequentItemsetsEU)

11

In [ ]:
# หาสินค้าที่ถูกซื้อซ้ำบ่อยในทวีปยุโรป โดยกำหนดค่า min_sup = 0.07  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.03

frequentItem_EU = list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItem_EU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItem_EU

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

[RelationRecord(items=frozenset({'20712'}), support=0.03365590144184058, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.03365590144184058, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.032567719870972754, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.032567719870972754, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04061249076988846, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04061249076988846, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.03999067272939256, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset(

output
[RelationRecord(items=frozenset({'22423'}), support=0.08355679919163654,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08355679919163654, lift=1.0)]),

RelationRecord(items=frozenset({'85099B'}), support=0.0825852085033617, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0825852085033617, lift=1.0)]),

RelationRecord(items=frozenset({'85123A'}), support=0.08713225292448797, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08713225292448797, lift=1.0)])]

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423')| (data['StockCode']=='85123A')]['Description'])

{'?',
 'CREAM HANGING HEART T-LIGHT HOLDER',
 'REGENCY CAKESTAND 3 TIER',
 'WHITE HANGING HEART T-LIGHT HOLDER',
 'damages',
 'faulty',
 'wrongly marked carton 22804'}

#ข้อ3 หาข้อมูลของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

### จัดการข้อมูลในช่วงตรุษจีนของทวีปเอเชีย

In [ ]:
Asia = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [ ]:
ASIA_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  ASIA_data=pd.concat([x,ASIA_data])
ASIA_data

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
423776    573154     22452        MEASURING TAPE BABUSHKA PINK        30   
423777    573154     22492             MINI PAINT SET VINTAGE         36   
423778    573154     23328  SET 6 SCHOOL MILK BOTTLES IN CRATE         8   
465423   C576233         M                              Manual        -1   
465424    576234         M                              Manual         1   

               InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571  2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572  2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
...                    ...        ...         ...                   ...   
423776 2011-10-28 08:20:00       2.95         NaN             Hong Kong   
423777 2011-10-28 08:20:00       0.65         NaN             Hong Kong   
423778 2011-10-28 08:20:00       3.75         NaN             Hong Kong   
465423 2011-11-14 13:26:00     326.10         NaN             Hong Kong   
465424 2011-11-14 13:27:00     326.10         NaN             Hong Kong   

              Date  month  day  
89570   2011-02-14      2   14  
89571   2011-02-14      2   14  
89572   2011-02-14      2   14  
89573   2011-02-14      2   14  
89574   2011-02-14      2   14  
...            ...    ...  ...  
423776  2011-10-28     10   28  
423777  2011-10-28     10   28  
423778  2011-10-28     10   28  
465423  2011-11-14     11   14  
465424  2011-11-14     11   14  

[1314 rows x 11 columns]

In [ ]:
As_data=ASIA_data[(ASIA_data['month']==1)]
As_data.head()

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   

              InvoiceDate  UnitPrice  CustomerID    Country        Date  \
70758 2011-01-25 13:26:00       0.36     12744.0  Singapore  2011-01-25   
70759 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70760 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70761 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
70762 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   

       month  day  
70758      1   25  
70759      1   25  
70760      1   25  
70761      1   25  
70762      1   25

In [ ]:
As_data_d=As_data[(As_data['day']>=15)]
As_data_d.head()

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   

              InvoiceDate  UnitPrice  CustomerID    Country        Date  \
70758 2011-01-25 13:26:00       0.36     12744.0  Singapore  2011-01-25   
70759 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70760 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70761 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
70762 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   

       month  day  
70758      1   25  
70759      1   25  
70760      1   25  
70761      1   25  
70762      1   25

In [ ]:
As_data2=ASIA_data[(ASIA_data['month']==2)]
As_data2.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [ ]:
As_data_d2=As_data2[(As_data2['day']<=15)]
As_data_d2.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [ ]:
# รวมทั้ง 2 เดือนเข้าด้วยกัน ในช่วงตรุษจีน
Asia_data_chi=pd.concat([As_data_d,As_data_d2])
Asia_data_chi

InvoiceNo StockCode                          Description  Quantity  \
58345    541238     21115                ROSE CARAVAN DOORSTOP         1   
58346    541238     22957  SET 3 PAPER VINTAGE CHICK PAPER EGG         1   
58347    541238     22758     LARGE PURPLE BABUSHKA NOTEBOOK           1   
58348    541238     20750             RED RETROSPOT MINI CASES         2   
58349    541238     22722    SET OF 6 SPICE TINS PANTRY DESIGN         1   
...         ...       ...                                  ...       ...   
80224    543027     22551             PLASTERS IN TIN SPACEBOY        12   
80225    543027     22557     PLASTERS IN TIN VINTAGE PAISLEY         12   
80226    543027     20974          12 PENCILS SMALL TUBE SKULL        24   
80227    543027     20975  12 PENCILS SMALL TUBE RED RETROSPOT        48   
80228    543027      POST                              POSTAGE         1   

              InvoiceDate  UnitPrice  CustomerID         Country        Date  \
58345 2011-01-16 10:42:00       6.75     16746.0  United Kingdom  2011-01-16   
58346 2011-01-16 10:42:00       2.95     16746.0  United Kingdom  2011-01-16   
58347 2011-01-16 10:42:00       1.25     16746.0  United Kingdom  2011-01-16   
58348 2011-01-16 10:42:00       7.95     16746.0  United Kingdom  2011-01-16   
58349 2011-01-16 10:42:00       3.95     16746.0  United Kingdom  2011-01-16   
...                   ...        ...         ...             ...         ...   
80224 2011-02-02 14:51:00       1.65     12414.0         Austria  2011-02-02   
80225 2011-02-02 14:51:00       1.65     12414.0         Austria  2011-02-02   
80226 2011-02-02 14:51:00       0.65     12414.0         Austria  2011-02-02   
80227 2011-02-02 14:51:00       0.65     12414.0         Austria  2011-02-02   
80228 2011-02-02 14:51:00      40.00     12414.0         Austria  2011-02-02   

       month  day  
58345      1   16  
58346      1   16  
58347      1   16  
58348      1   16  
58349      1   16  
...      ...  ...  
80224      2    2  
80225      2    2  
80226      2    2  
80227      2    2  
80228      2    2  

[32565 rows x 11 columns]

### หาข้อมูล

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data
transactionsAsiaChi = Asia_data_chi.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
min_sup = 0.082
frequentItemsets_Asia= list(apriori(transactionsAsiaChi,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsets_Asia

[RelationRecord(items=frozenset({'22423'}), support=0.08358591746327815, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08358591746327815, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08257558094349887, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08257558094349887, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08712209528250564, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08712209528250564, lift=1.0)])]

[RelationRecord(items=frozenset({'22423'}), support=0.08432400502104408, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08432400502104408, lift=1.0)]),

 RelationRecord(items=frozenset({'85099B'}), support=0.0823303551650299, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0823303551650299, lift=1.0)]),


 RelationRecord(items=frozenset({'85123A'}), support=0.08993576017130621, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08993576017130621, lift=1.0)])]

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423') | (data['StockCode']=='85099B')| (data['StockCode']=='85123A')]['Description'])

{'?',
 'CREAM HANGING HEART T-LIGHT HOLDER',
 'JUMBO BAG RED RETROSPOT',
 'REGENCY CAKESTAND 3 TIER',
 'WHITE HANGING HEART T-LIGHT HOLDER',
 'damages',
 'faulty',
 'wrongly marked carton 22804'}

# ข้อ 4 ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)

In [ ]:
Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [ ]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
536964 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536965 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536966 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536967 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536968 2011-12-08 10:26:00      40.00     12358.0         Austria  2011-12-08   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
536964     12    8  
536965     12    8  
536966     12    8  
536967     12    8  
536968     12    8  

[538358 rows x 11 columns]

In [ ]:
Eu_data=EU_data[(EU_data['month']==12)]
Eu_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   

   month  day  
0     12    1  
1     12    1  
2     12    1  
3     12    1  
4     12    1

In [ ]:
Eu_data_d=Eu_data[(Eu_data['day']<=31)]
Eu_data_d.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   

   month  day  
0     12    1  
1     12    1  
2     12    1  
3     12    1  
4     12    1

## หาข้อมูล

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data
transactionsEU_C_N = Eu_data_d.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
min_sup =0.08
frequentItemsetsEU= list(apriori(transactionsEU_C_N,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsetsEU

[RelationRecord(items=frozenset({'22086'}), support=0.09283118599273207, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09283118599273207, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08259002312520647, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08259002312520647, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.0981169474727453, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.0981169474727453, lift=1.0)])]

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22086')|(data['StockCode']=='22423')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS ",
 'REGENCY CAKESTAND 3 TIER',
 'damages',
 'faulty'}

# ข้อ 5 ของวันวาเลนไทน์ (14 Feb)

In [ ]:
Valen_data = data[(data['month']==2)]
Valen_data

InvoiceNo StockCode                        Description  Quantity  \
77628     542776     21671       RED SPOT CERAMIC DRAWER KNOB        48   
77629     542776     21668     RED STRIPE CERAMIC DRAWER KNOB        12   
77630     542776     21670      BLUE SPOT CERAMIC DRAWER KNOB        12   
77631     542776    17013D       ORIGAMI ROSE INCENSE IN TUBE        96   
77632     542776     17096  ASSORTED LAQUERED INCENSE HOLDERS        48   
...          ...       ...                                ...       ...   
105330    545218     21936           RED RETROSPOT PICNIC BAG         5   
105331    545218     21932    SCANDINAVIAN PAISLEY PICNIC BAG         5   
105332    545218     22386            JUMBO BAG PINK POLKADOT        10   
105333    545218     21933    PINK VINTAGE PAISLEY PICNIC BAG         5   
105334    545219     20738                                NaN       -36   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
77628  2011-02-01 08:23:00       1.25     15240.0  United Kingdom  2011-02-01   
77629  2011-02-01 08:23:00       1.25     15240.0  United Kingdom  2011-02-01   
77630  2011-02-01 08:23:00       1.25     15240.0  United Kingdom  2011-02-01   
77631  2011-02-01 08:23:00       0.42     15240.0  United Kingdom  2011-02-01   
77632  2011-02-01 08:23:00       0.21     15240.0  United Kingdom  2011-02-01   
...                    ...        ...         ...             ...         ...   
105330 2011-02-28 17:01:00       2.95     17462.0  United Kingdom  2011-02-28   
105331 2011-02-28 17:01:00       2.95     17462.0  United Kingdom  2011-02-28   
105332 2011-02-28 17:01:00       1.95     17462.0  United Kingdom  2011-02-28   
105333 2011-02-28 17:01:00       2.95     17462.0  United Kingdom  2011-02-28   
105334 2011-02-28 17:04:00       0.00         NaN  United Kingdom  2011-02-28   

        month  day  
77628       2    1  
77629       2    1  
77630       2    1  
77631       2    1  
77632       2    1  
...       ...  ...  
105330      2   28  
105331      2   28  
105332      2   28  
105333      2   28  
105334      2   28  

[27707 rows x 11 columns]

In [ ]:
valen_day=Valen_data[Valen_data['day']==14]
valen_day

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country        Date  \
88847 2011-02-14 09:24:00       5.95     17667.0  United Kingdom  2011-02-14   
88848 2011-02-14 09:24:00       5.95     17667.0  United Kingdom  2011-02-14   
88849 2011-02-14 09:24:00       5.95     17667.0  United Kingdom  2011-02-14   
88850 2011-02-14 09:24:00       0.42     17667.0  United Kingdom  2011-02-14   
88851 2011-02-14 09:24:00       4.95     17667.0  United Kingdom  2011-02-14   
...                   ...        ...         ...             ...         ...   
89950 2011-02-14 15:43:00       1.25     14606.0  United Kingdom  2011-02-14   
89951 2011-02-14 15:43:00       0.85     14606.0  United Kingdom  2011-02-14   
89952 2011-02-14 15:43:00       1.45     14606.0  United Kingdom  2011-02-14   
89953 2011-02-14 15:43:00       6.95     14606.0  United Kingdom  2011-02-14   
89954 2011-02-14 16:39:00       3.75     14189.0  United Kingdom  2011-02-14   

       month  day  
88847      2   14  
88848      2   14  
88849      2   14  
88850      2   14  
88851      2   14  
...      ...  ...  
89950      2   14  
89951      2   14  
89952      2   14  
89953      2   14  
89954      2   14  

[1108 rows x 11 columns]

In [ ]:
# จัดกลุ่มตาม transactions

transactions_valen = valen_day.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
# นับจำนวนสินค้าในช่วงวาเลนไทน์

len(transactions_valen)

55

In [ ]:
min_sup = 0.09
frequentItemsets_valen= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_valen:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_valen

yes
RelationRecord(items=frozenset({'22722', '22720'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22722', '22720'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])
yes
RelationRecord(items=frozenset({'22720', '22969'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22969'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.625, lift=6.875), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.875)])


[RelationRecord(items=frozenset({'20725'}), support=0.10909090909090909, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.10909090909090909, lift=1.0)]),
 RelationRecord(items=frozenset({'21621'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21621'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22383'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22411'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22411'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

# ข้อ 6 จัดกลุ่มตามประเทศ + เดือน โดยไม่มีเงื่อนไข

In [ ]:
transactions_M_C = data.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_M_C

Country      month
Australia    1        [22078, 22079, 22080, 22077, 22505, 22516, 225...
             2        [84978, 22960, 47590B, 47566, 22087, 22720, 22...
             3        [22978, 22980, 22993, 22654, 22968, 22505, 225...
             4        [POST, 22232, 22993, 22449, 20726, 20725, 2072...
             5        [22699, 22859, 22730, 84663A, 23206, 48129, 23...
                                            ...                        
Unspecified  6        [23122, 23119, 23108, 23240, 23237, 23239, 232...
             7        [21634, 21212, 22964, 23254, 23256, 84378, 209...
             8        [23236, 21210, 23313, 23318, 22952, 22950, 232...
             9        [22904, 21329, 21992, 20772, 23196, 22543, 225...
             11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
Name: StockCode, Length: 296, dtype: object

In [ ]:
min_sup = 0.32
frequentItemsets_M_C = list(apriori(transactions_M_C,min_support = min_sup))
for i in frequentItemsets_M_C:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_M_C

yes
RelationRecord(items=frozenset({'22326', 'POST'}), support=0.32094594594594594, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326', 'POST'}), confidence=0.32094594594594594, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8407079646017699, lift=1.4638209266007287), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5588235294117647, lift=1.463820926600729)])


[RelationRecord(items=frozenset({'20725'}), support=0.3547297297297297, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.3547297297297297, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.3716216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.3716216216216216, lift=1.0)]),
 RelationRecord(items=frozenset({'22139'}), support=0.3277027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22139'}), confidence=0.3277027027027027, lift=1.0)]),
 RelationRecord(items=frozenset({'22326'}), support=0.38175675675675674, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326'}), confidence=0.38175675675675674, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.4594594594594595, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326') | (data['StockCode']=='POST')]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

##ข้อ 6.1 group by ตามเดือนและประเทศ ในช่วงเวลาคริสต์มาสและปีใหม่

In [ ]:
transactions_C_N = Christm_Newy.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_C_N

Country          month
Australia        12       [22941, 21622, 21791, 35004C, 35004G, 85014B, ...
Austria          12       [22153, 37449, 37446, 22962, 21428, 22113, 223...
Bahrain          12                                                [72802B]
Belgium          12       [84375, 21217, 21212, 21977, 22417, 21975, 219...
Channel Islands  12       [22690, 21955, 20685, 21523, 21169, 22112, 214...
Cyprus           12       [85123A, 21031, 21026, 21025, 21032, 21030, 21...
Denmark          12       [22847, 22779, 22331, 22365, 22673, 22670, 825...
EIRE             12       [22968, 85071A, 85071C, 22355, 21579, 21576, 2...
Finland          12       [84992, 22951, 84991, 84991, 21213, 21213, 226...
France           12       [22728, 22727, 22726, 21724, 21883, 10002, 217...
Germany          12       [22809, 84347, 84945, 22242, 22244, 22243, 474...
Greece           12       [POST, 20685, 22941, 22816, 22966, 20725, 2072...
Iceland          12       [85116, 22375, 71477, 22492, 22771, 227

In [ ]:
min_sup =0.3
frequentItemsets_C_N = list(apriori(transactions_C_N,min_support = min_sup))
for i in frequentItemsets_C_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_C_N

yes
RelationRecord(items=frozenset({'POST', '20750'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '20750'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20750'}), items_add=frozenset({'POST'}), confidence=0.9, lift=1.7357142857142858), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'20750'}), confidence=0.6428571428571429, lift=1.735714285714286)])
yes
RelationRecord(items=frozenset({'22727', '22728'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22727', '22728'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'22727'}), items_add=frozenset({'22728'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22728'}), items_add=frozenset({'22727'}), confidence=1.0, lift=3.0)])


[RelationRecord(items=frozenset({'20685'}), support=0.37037037037037035, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20685'}), confidence=0.37037037037037035, lift=1.0)]),
 RelationRecord(items=frozenset({'20750'}), support=0.37037037037037035, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20750'}), confidence=0.37037037037037035, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.3333333333333333, lift=1.0)]),
 RelationRecord(items=frozenset({'21731'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21731'}), confidence=0.3333333333333333, lift=1.0)]),
 RelationRecord(items=frozenset({'21977'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21977

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='POST')|(data['StockCode']=='20750')]['Description'])

{'POSTAGE', 'RED RETROSPOT MINI CASES', nan}

##ข้อ 6.2 group by ตามเดือนและประเทศ ในทวีปยุโรป

In [ ]:
transactionsEU = EU_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.4
frequentItemsetsEU= list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsetsEU

yes
RelationRecord(items=frozenset({'22326', 'POST'}), support=0.41150442477876104, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326', 'POST'}), confidence=0.41150442477876104, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.9207920792079208, lift=1.2688964018353055), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5670731707317073, lift=1.2688964018353055)])


[RelationRecord(items=frozenset({'20725'}), support=0.4026548672566372, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.4026548672566372, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.45132743362831856, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.45132743362831856, lift=1.0)]),
 RelationRecord(items=frozenset({'22326'}), support=0.4469026548672566, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326'}), confidence=0.4469026548672566, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.5132743362831859, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.5132743362831859, lift=1.0)]),
 RelationRecord(items=frozenset({'22554'}), support=0.415929203539823, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22554'})

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

##ข้อ 6.3 group by ตามเดือนและประเทศ ในช่วงเวลาตรุษจีนในทวีเอเชีย

In [ ]:
transactionsAsia = Asia_data_chi.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:
min_sup = 0.35
frequentItemsets_Asia= list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_Asia

yes
RelationRecord(items=frozenset({'22326', 'POST'}), support=0.375, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326', 'POST'}), confidence=0.375, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.9230769230769231, lift=1.476923076923077), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.6, lift=1.4769230769230768)])


[RelationRecord(items=frozenset({'20726'}), support=0.375, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20726'}), confidence=0.375, lift=1.0)]),
 RelationRecord(items=frozenset({'22326'}), support=0.40625, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326'}), confidence=0.40625, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.46875, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.46875, lift=1.0)]),
 RelationRecord(items=frozenset({'22720'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720'}), confidence=0.5, lift=1.0)]),
 RelationRecord(items=frozenset({'POST'}), support=0.625, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST'}), confidence=0.625, lift=1.0)]),
 RelationRecord(items=frozenset({'22326', 'POST'}), support=0.375, ordered_statisti

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

## ข้อ 6.4  group by ตามเดือนและประเทศ ในช่วงคริสต์มาสและปีใหม่ในยุโรป

In [ ]:
transactionsEU_C_N = EU_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.4
frequentItemsetsEU= list(apriori(transactionsEU_C_N,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsetsEU

yes
RelationRecord(items=frozenset({'22326', 'POST'}), support=0.41150442477876104, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326', 'POST'}), confidence=0.41150442477876104, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.9207920792079208, lift=1.2688964018353055), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5670731707317073, lift=1.2688964018353055)])


[RelationRecord(items=frozenset({'20725'}), support=0.4026548672566372, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.4026548672566372, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.45132743362831856, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.45132743362831856, lift=1.0)]),
 RelationRecord(items=frozenset({'22326'}), support=0.4469026548672566, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326'}), confidence=0.4469026548672566, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.5132743362831859, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.5132743362831859, lift=1.0)]),
 RelationRecord(items=frozenset({'22554'}), support=0.415929203539823, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22554'})

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

## ข้อ 6.5  group by ตามเดือนและประเทศ ในช่วงวาเลนไทน์

In [ ]:
transactions_va = va_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.08
frequentItemsets_valen= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_valen:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_valen

yes
RelationRecord(items=frozenset({'22722', '22720'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22722', '22720'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])
yes
RelationRecord(items=frozenset({'22720', '22969'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22969'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.625, lift=6.875), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.875)])


[RelationRecord(items=frozenset({'20725'}), support=0.10909090909090909, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.10909090909090909, lift=1.0)]),
 RelationRecord(items=frozenset({'21621'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21621'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22383'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22411'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22411'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22722')|(data['StockCode']=='22720')]['Description'])

{'SET OF 3 CAKE TINS PANTRY DESIGN ', 'SET OF 6 SPICE TINS PANTRY DESIGN', nan}